lib/irb/type_completion/types.rb

```diff
@@ -0,0 +1,443 @@
+                expected << rest.type
+              end
+              if given.any?
+                score += given.zip(expected).count do |t, e|
+                  e = from_rbs_type e, receiver_type
+                  intersect?(t, e) || (intersect?(STRING, e) && t.methods.include?(:to_str)) || (intersect?(INTEGER, e) && t.methods.include?(:to_int)) || (intersect?(ARRAY, e) && t.methods.include?(:to_ary))
+                end.fdiv(given.size)
+              end
+            end
+            [[method_type, given || [], expected || []], score]
+          end
+        end
+        max_score = methods_with_score.map(&:last).max
+        methods_with_score.select { _2 == max_score }.map(&:first)
+      end
+
+      def self.intersect?(a, b)
+        atypes = a.types.group_by(&:class)
+        btypes = b.types.group_by(&:class)
+        if atypes[SingletonType] && btypes[SingletonType]
+          aa, bb = [atypes, btypes].map {|types| types[SingletonType].map(&:module_or_class) }
+          return true if (aa & bb).any?
+        end
+
+        aa, bb = [atypes, btypes].map {|types| (types[InstanceType] || []).map(&:klass) }
+        (aa.flat_map(&:ancestors) & bb).any?
+      end
+
+      def self.type_from_object(object)
+        case object
+        when Array, Hash, Module
+          type_from_object_recursive(object, max_level: 4)
+        else
+          klass = object.singleton_class rescue object.class
+          InstanceType.new klass
+        end
+      end
+
+      def self.type_from_object_recursive(object, max_level:)
+        max_level -= 1
+        sample_size = 1000
+        case object
+        when Array
+          values = object.size > sample_size ? object.sample(sample_size) : object
+          if max_level > 0
+            InstanceType.new Array, { Elem: UnionType[*values.map { type_from_object_recursive(_1, max_level: max_level) }] }
+          else
+            InstanceType.new Array, { Elem: UnionType[*values.map(&:class).uniq.map { InstanceType.new _1 }] }
+          end
+        when Hash
+          keys = object.size > sample_size ? object.keys.sample(sample_size) : object.keys
+          values = object.size > sample_size ? object.values.sample(sample_size) : object.values
+          if max_level > 0
+            key_types = keys.map { type_from_object_recursive(_1, max_level: max_level) }
+            value_types = values.map { type_from_object_recursive(_1, max_level: max_level) }
+            InstanceType.new Hash, { K: UnionType[*key_types], V: UnionType[*value_types] }
+          else
+            key_types = keys.map(&:class).uniq.map { InstanceType.new _1 }
+            value_types = values.map(&:class).uniq.map { InstanceType.new _1 }
+            InstanceType.new Hash, { K: UnionType[*key_types], V: UnionType[*value_types] }
+          end
+        when Module
+          SingletonType.new object
+        else
+          InstanceType.new object.class
+        end
```

Comment:
I wonder if we should have a rescue here as a circuit breaker and log the exception to help issue reporting?

Reply:
Added test for BasicObject. `Types.type_from_object` is not wrapped with begin-rescue with fallback type anymore. So now exception will be caught and stored to `Completor.last_completion_error` for issue reporting.